cas-server-core-authentication/src/main/java/org/jasig/cas/authentication/DefaultAuthenticationContextBuilder.java

```diff
@@ -0,0 +1,177 @@
+    @Override
+    public boolean collect(final Authentication authentication) throws AuthenticationException {
+        if (this.authentications.add(authentication)) {
+            LOGGER.debug("Collected authentication event. Associated principal with this authentication is [{}]",
+                    authentication.getPrincipal());
+            return true;
+        }
+        LOGGER.warn("Failed to collect duplicate authentication event as it already exists.");
+        return false;
+    }
+
+    @Override
+    public AuthenticationContext build() {
+        final Authentication authentication = buildAuthentication();
+        if (authentication == null) {
+            throw new RuntimeException(new GeneralSecurityException("Authentication context cannot be produced"));
+        }
+        return new DefaultAuthenticationContext(authentication);
+    }
+
+    @Override
+    public void clear() {
+        this.authentications.clear();
+    }
+
+    private Authentication buildAuthentication() {
+        if (isEmpty()) {
+            LOGGER.warn("No authentication event has been recorded; CAS cannot finalize the authentication context");
+            return null;
+        }
+
+        final Map<String, Object> authenticationAttributes = new HashMap<>();
+        final Map<String, Object> principalAttributes = new HashMap<>();
+        final AuthenticationBuilder authenticationBuilder = DefaultAuthenticationBuilder.newInstance();
+
+        buildAuthenticationHistory(authenticationAttributes, principalAttributes, authenticationBuilder);
+
+        final Principal primaryPrincipal = getPrimaryPrincipal(principalAttributes);
+        authenticationBuilder.setPrincipal(primaryPrincipal);
+        LOGGER.debug("Determined primary authentication principal to be [{}]", primaryPrincipal);
+
+        authenticationBuilder.setAttributes(authenticationAttributes);
+        LOGGER.debug("Collected authentication attributes for this context are [{}]", authenticationAttributes);
+
+        final DateTime dt = DateTime.now();
+        authenticationBuilder.setAuthenticationDate(dt);
+        LOGGER.debug("Authentication context commenced at [{}]", dt);
+
+        return authenticationBuilder.build();
+
+    }
+
+    private void buildAuthenticationHistory(final Map<String, Object> authenticationAttributes,
+                                            final Map<String, Object> principalAttributes,
+                                            final AuthenticationBuilder authenticationBuilder) {
+
+        LOGGER.debug("Collecting authentication history based on [{}] authentication events", this.authentications.size());
+        for (final Authentication authn : this.authentications) {
+            final Principal authenticatedPrincipal = authn.getPrincipal();
+            LOGGER.debug("Evaluating authentication principal [{}] for inclusion in context", authenticatedPrincipal);
+
+            principalAttributes.putAll(authenticatedPrincipal.getAttributes());
```

Comment:
Consider the following example:

1. Step 1, authenticate with username/password: Jérôme/password
2. Step 2, authenticate against stormpath

Your primary principal id is selected by the strategy to be: Jérôme 
Your principal attributes are a combination of step1 and step2. 
Your authentication attributes are also a combination of step1 and step2.

After all, your identity is a combination of step1 and step2. To the final outer layer, you are the entire event, because that is everything we know about you. 

In most deployments, step2 does not produce any meaningful attributes, but in some cases, you may want to learn a bit more about the principal; so we need to cover steps.